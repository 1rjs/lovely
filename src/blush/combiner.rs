use crate::{
    blush::printer,
    parser::ast::{Expression, ExpressionKind, Program, UseTailItem},
    span::Span,
};
use std::error::Error;

pub struct Combiner<'a> {
    files: Vec<(&'a str, &'a Program)>,
}

// TODO:
// - namespace imports
// - naming conflicts (might have to mangle imported variable names)

impl<'a> Combiner<'a> {
    pub const fn new(programs: Vec<(&'a str, &'a Program)>) -> Self {
        Self { files: programs }
    }

    pub fn combine(self) -> Result<Program, Box<dyn Error>> {
        printer::info(&format!("Combining {} ASTs...", self.files.len()));
        let Some(main_file) = self
            .files
            .iter()
            .find(|(name, _)| name.eq_ignore_ascii_case("main.lv"))
        else {
            return Err("No main.lv file found".into());
        };

        self.resolve_imports(main_file.1)
    }

    fn resolve_imports(&self, main_file: &Program) -> Result<Program, Box<dyn Error>> {
        let mut new_main_file = Program(vec![]);

        for expr in &main_file.0 {
            match &expr.kind {
                crate::parser::ast::ExpressionKind::Use { segments, tail } => {
                    let mut ref_program_name = segments.join("/");
                    ref_program_name.push_str(".lv");
                    let Some((_, ref_program)) = self
                        .files
                        .iter()
                        .find(|(name, _)| name == &ref_program_name)
                    else {
                        return Err(
                            format!("Couldn't find reference program {ref_program_name}").into(),
                        );
                    };

                    if tail.is_empty() {
                        todo!("namespaced imports")
                    } else {
                        for tail_item in tail {
                            let Some(corresponding_def) =
                                ref_program.0.iter().find(|expr| match &expr.kind {
                                    ExpressionKind::VariableDecl { name, .. } => {
                                        name == &tail_item.name
                                    }
                                    _ => false,
                                })
                            else {
                                return Err(format!(
                                    "Couldn't find corresponding definition for {tail_item:?}"
                                )
                                .into());
                            };

                            let referenced_vars =
                                Self::referenced_variables(corresponding_def, vec![]);

                            let mut uses_used_in_corresponding_def = ref_program
                                .0
                                .iter()
                                .filter_map(|expr| match &expr.kind {
                                    ExpressionKind::Use { segments, tail } => {
                                        Some((segments, tail))
                                    }
                                    _ => None,
                                })
                                .filter(|(segments, tail_items)| {
                                    referenced_vars.iter().any(|(name, namespace)| {
                                        Self::matches_import(
                                            name,
                                            namespace.as_ref(),
                                            segments,
                                            tail_items,
                                        )
                                    })
                                })
                                .map(|(segments, tail_items)| {
                                    Expression::new(
                                        ExpressionKind::Use {
                                            segments: segments.clone(),
                                            tail: tail_items.clone(),
                                        },
                                        Span::from_range(0, 0), // TODO: real span
                                    )
                                })
                                .collect::<Vec<_>>();

                            new_main_file.0.append(&mut uses_used_in_corresponding_def);
                            new_main_file.0.push(corresponding_def.clone());
                        }
                    }
                }

                ExpressionKind::BoolLiteral(_)
                | ExpressionKind::IntLiteral(_)
                | ExpressionKind::Ident { .. }
                | ExpressionKind::Block(_)
                | ExpressionKind::Prefix { .. }
                | ExpressionKind::Infix { .. }
                | ExpressionKind::VariableDecl { .. }
                | ExpressionKind::Function { .. }
                | ExpressionKind::FunctionCall { .. } => {
                    // TODO: some of these will need to be handled differently, like namespaced
                    // variable idents
                    new_main_file.0.push(expr.clone());
                }
            }
        }

        if new_main_file
            .0
            .iter()
            .any(|expr| matches!(&expr.kind, ExpressionKind::Use { .. }))
        {
            self.resolve_imports(&new_main_file)
        } else {
            Ok(new_main_file)
        }
    }

    fn referenced_variables(
        expr: &Expression,
        base: Vec<(String, Option<String>)>,
    ) -> Vec<(String, Option<String>)> {
        match &expr.kind {
            ExpressionKind::BoolLiteral(_) | ExpressionKind::IntLiteral(_) => base,
            ExpressionKind::Ident { name, namespace } => {
                let mut new_base = base;
                new_base.push((name.clone(), namespace.clone()));
                new_base
            }
            ExpressionKind::Block(expressions) => expressions
                .iter()
                .fold(base, |acc, expr| Self::referenced_variables(expr, acc)),
            ExpressionKind::Prefix { expression, .. } => {
                Self::referenced_variables(expression, base)
            }
            ExpressionKind::Infix { left, right, .. } => {
                let mut new_base = Self::referenced_variables(left, base);
                new_base = Self::referenced_variables(right, new_base);
                new_base
            }
            ExpressionKind::VariableDecl { value, .. } => Self::referenced_variables(value, base),
            ExpressionKind::Function { body, .. } => Self::referenced_variables(body, base),
            ExpressionKind::FunctionCall {
                name,
                namespace,
                arguments,
            } => {
                let mut new_base = base;
                new_base.push((name.clone(), namespace.clone()));
                arguments.iter().fold(new_base, |acc, arg| {
                    Self::referenced_variables(&arg.value, acc)
                })
            }
            ExpressionKind::Use { .. } => todo!("not sure what to do here yet"),
        }
    }

    fn matches_import(
        name: &str,
        namespace: Option<&String>,
        segments: &[String],
        tail_items: &[UseTailItem],
    ) -> bool {
        namespace.as_ref().map_or_else(
            || {
                tail_items.iter().any(|tail_item| {
                    tail_item
                        .alias
                        .as_ref()
                        .map_or_else(|| tail_item.name == name, |alias| alias == name)
                })
            },
            |namespace| tail_items.is_empty() && segments.last() == Some(namespace),
        )
    }
}

#[cfg(test)]
mod tests {
    use crate::{blush::combiner::Combiner, parser::ast::UseTailItem};

    #[test]
    fn test_matches_import() {
        // normal variable import
        assert!(Combiner::matches_import(
            "foo",
            None,
            &["lorem".to_string(), "ipsum".to_string()],
            &[UseTailItem {
                name: "foo".to_string(),
                alias: None
            }]
        ));

        // missing variable import
        assert!(!Combiner::matches_import(
            "foo",
            None,
            &["lorem".to_string(), "ipsum".to_string()],
            &[]
        ));

        // aliased variable import
        assert!(Combiner::matches_import(
            "bar",
            None,
            &["lorem".to_string(), "ipsum".to_string()],
            &[UseTailItem {
                name: "foo".to_string(),
                alias: Some("bar".to_string())
            }]
        ));

        // aliased but using non-aliased name should fail
        assert!(!Combiner::matches_import(
            "foo",
            None,
            &["lorem".to_string(), "ipsum".to_string()],
            &[UseTailItem {
                name: "foo".to_string(),
                alias: Some("bar".to_string())
            }]
        ));

        // namespaced import
        assert!(Combiner::matches_import(
            "foo",
            Some(&"ipsum".to_string()),
            &["lorem".to_string(), "ipsum".to_string()],
            &[]
        ));
    }
}
