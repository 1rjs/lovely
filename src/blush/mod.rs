// lovingly named by japi (July 24, 2025)

use crate::codegen::emitters::Emitter;
use crate::{
    blush::combiner::Combiner, checker::Checker,
    codegen::emitters::x86_64_linux_nasm::CodeGenerator, ir::IRGenerator, parser::Parser,
};
use std::{
    error::Error,
    fs,
    path::{Path, PathBuf},
};
use toml::Table;

pub mod combiner;
pub mod printer;

pub struct Blush {}

impl Blush {
    pub fn build(path: String) -> Result<(), Box<dyn Error>> {
        let path_buf = PathBuf::from(path);
        let file_tree = collect_files(&path_buf)?;

        printer::info("Reading config file...");
        let Some(config_file) = file_tree.find(&|name, _| {
            name.split('/')
                .next_back()
                .expect("No file name")
                .eq_ignore_ascii_case("config.toml")
        }) else {
            return Err("No config file found".into());
        };
        let Ok(config) = config_file.parse::<Table>() else {
            return Err("Failed to parse config file".into());
        };
        let Some(Some(package_name)) = config.get("package").map(|s| s.as_str()) else {
            return Err("No package name found in config file".into());
        };

        let lovely_files = file_tree
            .filter(&|name, _| {
                std::path::Path::new(name)
                    .extension()
                    .is_some_and(|ext| ext.eq_ignore_ascii_case("lv"))
            })
            .expect("Expected a directory");

        printer::info(&format!("Parsing {} files...", lovely_files.len()));
        let parsed_lovely_files =
            lovely_files.try_map(&|s| Parser::new(s).parse().map_err(std::convert::Into::into))?;
        let flattened_files: Vec<(String, &_)> = parsed_lovely_files
            .flatten()
            .iter()
            .map(|(name, program)| (trim_path(name, package_name), *program))
            .collect();

        let combined_program = Combiner::new(
            flattened_files
                .iter()
                .map(|(name, program)| (name.as_str(), *program))
                .collect(),
        )
        .combine()?;

        let mut checker = Checker::new();
        let Ok(checked_program) = checker.check(&combined_program) else {
            return Err("Failed to typecheck program".into());
        };

        let ir = IRGenerator::new(checker.types).program_ir(&checked_program);
        let asm = CodeGenerator::new().gen_asm(&ir);

        println!("\n---\n{asm}\n---\n");

        Ok(())
    }
}

fn collect_files(path: &Path) -> Result<FileTreeNode<String>, Box<dyn Error>> {
    if !path.is_dir() {
        printer::error("That's not a directory you silly!");
        return Err("Not a directory".into());
    }
    let Ok(dir) = path.read_dir() else {
        printer::error("Uh for some reason we couldn't read the directory. Not really sure why.");
        return Err("Failed to read directory".into());
    };
    let Ok(dir) = dir.collect::<Result<Vec<_>, _>>() else {
        printer::error(
            "You know, today just really isn't your day is it. Something awful happened, and I have no idea what it was.",
        );
        return Err("Failed to read directory".into());
    };

    let mut children = vec![];

    for dir_entry in dir {
        if dir_entry.path().is_dir() {
            children.push(collect_files(&dir_entry.path())?);
        } else {
            let contents = fs::read_to_string(dir_entry.path())?;
            let name = dir_entry.path().to_str().unwrap().to_string();
            children.push(FileTreeNode::File { name, contents });
        }
    }

    Ok(FileTreeNode::Directory {
        children,
        name: path
            .to_str()
            .expect("Weird file name, it made me uncomfortable.")
            .into(),
    })
}

enum FileTreeNode<T> {
    Directory {
        name: String,
        children: Vec<FileTreeNode<T>>,
    },
    File {
        name: String,
        contents: T,
    },
}

impl<T> FileTreeNode<T> {
    fn filter<F>(&self, f: &F) -> Option<Self>
    where
        F: Fn(&String, &T) -> bool,
        T: Clone,
    {
        match self {
            Self::Directory { children, name } => Some(Self::Directory {
                name: name.clone(),
                children: children.iter().filter_map(|c| c.filter(f)).collect(),
            }),
            Self::File { name, contents } => {
                if f(name, contents) {
                    Some(Self::File {
                        name: name.clone(),
                        contents: contents.clone(),
                    })
                } else {
                    None
                }
            }
        }
    }

    fn map<U>(&self, f: &impl Fn(&T) -> U) -> FileTreeNode<U> {
        match self {
            Self::Directory { name, children } => FileTreeNode::Directory {
                name: name.clone(),
                children: children.iter().map(|c| Self::map(c, f)).collect(),
            },
            Self::File { name, contents } => FileTreeNode::File {
                name: name.clone(),
                contents: f(contents),
            },
        }
    }

    fn try_map<U>(
        &self,
        f: &impl Fn(&T) -> Result<U, Box<dyn Error>>,
    ) -> Result<FileTreeNode<U>, Box<dyn Error>> {
        match self {
            Self::Directory { name, children } => Ok(FileTreeNode::Directory {
                name: name.clone(),
                children: children
                    .iter()
                    .map(|c| c.try_map(f))
                    .collect::<Result<Vec<_>, _>>()?,
            }),
            Self::File { name, contents } => Ok(FileTreeNode::File {
                name: name.clone(),
                contents: f(contents)?,
            }),
        }
    }

    fn find(&self, f: &impl Fn(&String, &T) -> bool) -> Option<&T> {
        match self {
            Self::Directory { children, .. } => children.iter().find_map(|c| c.find(f)),
            Self::File { contents, name } => {
                if f(name, contents) {
                    Some(contents)
                } else {
                    None
                }
            }
        }
    }

    fn flatten(&self) -> Vec<(&str, &T)> {
        match self {
            Self::Directory { children, .. } => children.iter().flat_map(|c| c.flatten()).collect(),
            Self::File { contents, name } => vec![(name, contents)],
        }
    }

    fn len(&self) -> usize {
        match self {
            Self::Directory { children, .. } => children.iter().map(Self::len).sum(),
            Self::File { .. } => 1,
        }
    }
}

fn trim_path(file_path: &str, package_name: &str) -> String {
    let segments = file_path.split('/').collect::<Vec<_>>();
    let mut new_segments = Vec::new();

    let inside_dir = !segments.contains(&package_name);

    let mut has_hit_package_name = false;
    for segment in segments {
        if segment.eq_ignore_ascii_case(".") {
            continue;
        }
        if segment.eq_ignore_ascii_case(package_name) {
            has_hit_package_name = true;
            continue;
        }
        if has_hit_package_name || inside_dir {
            new_segments.push(segment);
        }
    }

    new_segments.join("/")
}

#[cfg(test)]
mod tests {
    use super::trim_path;

    #[test]
    fn trim_path_test() {
        assert_eq!(
            trim_path("my_package/foo/bar.lv", "my_package"),
            "foo/bar.lv"
        );
        assert_eq!(
            trim_path("./my_package/foo/bar.lv", "my_package"),
            "foo/bar.lv"
        );
        assert_eq!(
            trim_path(
                "/Users/sam/whatever/my_package/foo/bar/baz.lv",
                "my_package"
            ),
            "foo/bar/baz.lv"
        );

        // edge case for if you're in the directory...
        assert_eq!(
            trim_path("./foo/bar/baz.lv", "my_package"),
            "foo/bar/baz.lv"
        );
    }
}
