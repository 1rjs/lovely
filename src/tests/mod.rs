use crate::blush::Blush;
use std::fs;
use std::path::PathBuf;

fn get_test_case_dir_names() -> Vec<String> {
    let cases_dir = PathBuf::from("src/tests/cases");
    fs::read_dir(&cases_dir)
        .expect("Failed to read test cases directory")
        .filter_map(std::result::Result::ok)
        .map(|entry| entry.path())
        .filter(|path| path.is_dir())
        .map(|path| path.to_str().unwrap().to_string())
        .collect()
}

#[test]
fn snapshot_tests() {
    let test_case_dirs = get_test_case_dir_names();
    for dir_name in test_case_dirs {
        let asm = Blush::get_asm_and_package_name(&dir_name);
        insta::with_settings!({
            description => dir_name.clone(),
        }, {
            match asm {
                Ok((asm, _)) => {
                    insta::assert_snapshot!(asm);
                }
                Err(err) => {
                    insta::assert_snapshot!(err);
                }
            }
        });
    }
}
