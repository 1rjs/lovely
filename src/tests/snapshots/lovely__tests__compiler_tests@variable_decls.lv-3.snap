---
source: src/tests/mod.rs
expression: checked_program
input_file: src/tests/source_files/variable_decls.lv
---
CheckedProgram {
    stmts: [
        CheckedExpressionStatement {
            expr: CheckedExpression {
                type_id: 2,
                data: VariableDecl {
                    name: "foo",
                    value: CheckedExpression {
                        type_id: 0,
                        data: IntLiteral(
                            6,
                        ),
                    },
                    mutable: false,
                },
            },
        },
        CheckedExpressionStatement {
            expr: CheckedExpression {
                type_id: 2,
                data: VariableDecl {
                    name: "bar",
                    value: CheckedExpression {
                        type_id: 0,
                        data: IntLiteral(
                            7,
                        ),
                    },
                    mutable: true,
                },
            },
        },
        CheckedExpressionStatement {
            expr: CheckedExpression {
                type_id: 2,
                data: VariableDecl {
                    name: "baz",
                    value: CheckedExpression {
                        type_id: 1,
                        data: BoolLiteral(
                            true,
                        ),
                    },
                    mutable: true,
                },
            },
        },
        CheckedExpressionStatement {
            expr: CheckedExpression {
                type_id: 2,
                data: VariableDecl {
                    name: "qux",
                    value: CheckedExpression {
                        type_id: 2,
                        data: Unit,
                    },
                    mutable: false,
                },
            },
        },
    ],
}
